lesson-05-time-management/time.c

```diff
@@ -0,0 +1,181 @@
+	  split_time.tm_mday,
+	  split_time.tm_mon + 1,
+	  split_time.tm_year + 1900,
+	  split_time.tm_hour,
+	  split_time.tm_min,
+	  split_time.tm_sec);
+
+	do_gettimeofday(&prev_read_time);
+
+	return ret;
+}
+CLASS_ATTR_RO(prev_read_abs_time);
+
+/**
+ * Implement kernel module with API in sysfs or procfs, which return Fibonacci sequence. Each element of sequence should be generated once in a second.
+ */
+struct fibonacci_seq {
+	u64 prev_number;
+	u64 curr_number;
+	unsigned int seq_number;
+
+	spinlock_t        spin_lock;
+	struct timer_list timer;
+};
+struct fibonacci_seq fibonacci_seq;
+
+static void fibonacci_next(struct fibonacci_seq *seq) {
+	u64 prev;
+	unsigned long flags;
+
+	BUG_ON(NULL == seq);
+
+	spin_lock_irqsave(&seq->spin_lock, flags);
+
+	prev = seq->curr_number;
+	seq->curr_number = seq->curr_number + seq->prev_number;
+	seq->prev_number = prev;
+	seq->seq_number++;
+
+	spin_unlock_irqrestore(&seq->spin_lock, flags);
+}
+
+static u64 fibonacci_get_current(struct fibonacci_seq *seq, unsigned int *seq_number) {
+	unsigned int ret;
+	unsigned long flags;
+
+	spin_lock_irqsave(&seq->spin_lock, flags);
+
+	ret = seq->curr_number;
+	if (seq_number) {
+		*seq_number = seq->seq_number;
+	}
+
+	spin_unlock_irqrestore(&seq->spin_lock, flags);
+
+	return ret;
+}
+
+#define TIMER_PERIOD_JIFFIES (1 * HZ) /* 1 sec */
+
+static void fibonacci_timer_callback(unsigned long data) {
+	struct fibonacci_seq *seq = (struct fibonacci_seq *)data;
+
+	fibonacci_next(seq);
+	mod_timer(&seq->timer, get_jiffies_64() + (TIMER_PERIOD_JIFFIES));
+}
+
+static void fibonacci_timer_init(struct fibonacci_seq *seq) {
+	seq->prev_number = 0;
+	seq->curr_number = 1;
+	seq->seq_number  = 0;
+	spin_lock_init(&seq->spin_lock);
+
+	init_timer(&seq->timer);
+	setup_timer(&seq->timer, fibonacci_timer_callback, (unsigned int)seq);
+	mod_timer(&seq->timer, get_jiffies_64() + TIMER_PERIOD_JIFFIES);
+}
+
+static void fibonacci_timer_destroy(struct fibonacci_seq *seq) {
+	del_timer(&seq->timer);
+}
+
+static ssize_t fibonacci_seq_show(struct class *class, struct class_attribute *attr, char *buf) {
+	u64 current_fibonacci_number;
+	unsigned int current_seq_number;
+
+	current_fibonacci_number = fibonacci_get_current(&fibonacci_seq, &current_seq_number);
+
+	return scnprintf(buf, PAGE_SIZE, "seq: %u, value: %llu\n", current_seq_number, current_fibonacci_number);
+}
+CLASS_ATTR_RO(fibonacci_seq);
+
+static int __init time_init(void) {
+	int ret;
+
+	time_management_class = class_create(THIS_MODULE, "time-management");
+	if(IS_ERR(time_management_class)) {
+		printk(KERN_ERR "%s: bad class create\n", THIS_MODULE->name);
+		return -1;
+	}
+
+	ret = class_create_file(time_management_class, &class_attr_prev_read_rel_time);
+	if (0 != ret)
+		goto err_rel_time;
+
+	ret = class_create_file(time_management_class, &class_attr_prev_read_abs_time);
+	if (0 != ret)
+		goto err_abs_time;
+
+	ret = class_create_file(time_management_class, &class_attr_fibonacci_seq);
```

Comment:
Fibonacci functionality hasn't been initialized yet.
So `fibonacci_get_current()` can be accessed before structure (including lock) is initialized.